apache/livy
# [LIVY-616] Livy Server discovery

## What changes were proposed in this pull request?
JIRA ticket https://issues.apache.org/jira/browse/LIVY-616.
Implement Livy Server discovery mechanism based on ZooKeeper. And provide convenient API for clients to get LivyServer URL without explicitly set it manually by "livy.server.host" property. Then we also can use it in LivyClientBuilder to set URL automatically. It will provide a more easy way for configuration. We don't need to set "livy.server.host" or "zeppelin.livy.url" manually in a case when we don't know where Livy Server will be started.

I have separated ZooKeeper logic from StateStore logic, as a result, we can use new ZooKeeperManager just to communicate with ZooKeeper and build new logic based on it (as I did for DiscoveryManager and ZooKeeperStateStore)

## How was this patch tested?
Added unit tests for new functionality.
Tested manually:
After starting LivyServer we can get Livy Server URI from ZooKeeper
![image](https://user-images.githubusercontent.com/31073930/62357180-38035e00-b51b-11e9-9744-e3894c1acea8.png)

![image](https://user-images.githubusercontent.com/31073930/62357488-de4f6380-b51b-11e9-8ef0-eb0dbe549db2.png)
Also verified on a secured cluster.

What the best way to add new API?
I see the couple of ways to do that:
1. Add it to LivyConf - in this case, we need to create DiscoveryManager in LivyConf which is not good.
2. Add it to LivyClient interface and set some variable during creating client via LivyClientBuilder.

Please, advise.
